Cl-Cd (as in XRotor) fcn

CrescenzoVisone opened the pull request:
Creation of the function Lift Coefficient - Drag Coefficient according to XRotor aerodynamic model

CrescenzoVisone committed:
Branch and fcn creation

CrescenzoVisone committed:
First draft of the ClCd_XRotor.m fcn

CrescenzoVisone commented:
The first draft of the function, which evaluates the Cd(Cl) law according to XRotor aerodynamic model, is intended to take in input these values:
- (**Cd_min**) = minimum blade element's drag coefficient
- (**dCd_dCl2**) = quadratic coefficient of the parable that approximates the Cd(Cl) law
- (**Cl_Cd_min**) = blade element's lift coefficient for which the drag coefficient assumes its minimum value
- (**Re_ref**) = blade element's Reynolds number computed taking into account the radius at which the blade element is intended to be (i.e. the reference velocity is (_Omega*r_) where (_Omega_) is the propeller's angular velocity and (_r_) is the radial position of the blade element taken into account)
- (**Re_inf**) = asymptotic Reynolds number of the phenomena
- (**f**) = Reynolds number scaling exponent, according to XRotor documentation
- (**Cl_max**) = maximum blade element's lift coefficient
- (**Cl_min**) = minimum blade element's lift coefficient
- (**Cl_bp**) = blade element's lift coefficient breakpoints; this input can be both a vector or a single value; a breakpoint is the value at which the drag coefficient evaluation is requested, according to the Cd(Cl) law built through the function script

Apart from further corrections, future developments of this fcn could be:
**a.** taking input values in the form of a single vector, whose components are the above quantities; this could make the function calling more compact
**b.** insert a control algorithm to force the Reynolds number values and the Reynolds number scaling exponent to be coherent with each other

CrescenzoVisone committed:
Function header insertion

CrescenzoVisone committed:
Header corrections

CrescenzoVisone committed:
Changed input inserting mode

CrescenzoVisone commented:
In the second draft of the function, input data inserting mode has been changed: now the function takes 2 inputs;

- the first input is a vector whose components must be ordered in the following way:
         1.  **_Cd_min_**
         2.  **_dCd_dCl2_**
         3.  **_Cl_Cd_min_**
         4.  **_Re_ref_**
         5.  **_Re_inf_** 
         6.  **_f_**
         7.  **_Cl_max_**
         8.  _**Cl_min**_

(the meaning of each of these quantities is the same of the first draft and is specified in the previous comment)

- the second input can be either a vector or a single value, representing the blade element's lift coefficient breakpoints (i.e. it is the variable previously called **_Cl_bp_**)

### The code is now more compact and easier to read but the fcn must be carefully called because the order of the input vectors can only be the one just described. Input data given in a different order could result in wrong outputs.


The outputs of the fcn are:
a.  (_**Cl_vec**_) = this is the lift coefficients vector used to build the Cl-Cd law; it is through this vector that the parable is plotted
b.  (**_Cd_**) = this is the drag coefficients vector evaluated at _Cl_vec_
c.  (**_Cd_bp_**) = this is a vector or a single value, depending on the nature of the _Cl_bp_ input, whose components are the drag coefficients requested at lift coefficients breakpoints


Apart from further corrections, future developments of this fcn could be:
i. insert a control algorithm to force the Reynolds number values and the Reynolds number scaling exponent to be coherent with each other

EttoreSaetta commented on ClCd_XRotor.m:
```diff
@@ -0,0 +1,76 @@
+    % aerodynamic model building
+    for i = 1:length(Cl_vec)
+        Cl = Cl_vec(i);
+        Cd(i) = (Cd_min + dCd_dCl2*(Cl_Cd_min - Cl)^2)*(Re_inf/Re_ref)^f;
```
Here you could compute the Cd vector without the `for` cycle.
In matlab you can do a vector operation using a dot to specify that the operation is on the vector components.

EttoreSaetta commented on ClCd_XRotor.m:
```diff
@@ -0,0 +1,76 @@
+    % definition of an anonymous function through the handle (@) symbol
+    Cd_interp = @(Cl_interp) interp1(Cl_vec, Cd, Cl_interp, 'pchip');
+    % anonymous fcn used to compute requested Cd
+    Cd_bp = Cd_interp(Cl_bp);
```
It could be useful to plot also these values along the the non-dimensional radius for the same stations of the `Cl_bp` vector.

CrescenzoVisone replied:
The function does not take in input the non-dimensional radius stations of the Cl_bp.
Should I ask in input these data too?

EttoreSaetta replied:
Let's leave it like this for the moment, there are already many inputs.

CrescenzoVisone replied:
Ok, thank you

EttoreSaetta commented on ClCd_XRotor.m:
```diff
@@ -0,0 +1,76 @@
+        Cd(i) = (Cd_min + dCd_dCl2*(Cl_Cd_min - Cl)^2)*(Re_inf/Re_ref)^f;
+    end
+    % plot
+    plot(Cd, Cl_vec);
```
Put the plots at the end of the function and comment in a way that who has to do the output interface will understand.

EttoreSaetta reviewed (commented):
Here there is a first review of the function.
Just some improvements.

CrescenzoVisone committed:
Made corrections #1 and #3 suggested by @EttoreSaetta

CrescenzoVisone commented:
>

CrescenzoVisone committed:
TestCases folder creation

CrescenzoVisone committed:
Test script created for validation

CrescenzoVisone committed:
Created "doc" folder and added user guide documentation .pdf file

EttoreSaetta commented:
@CrescenzoVisone
If you name "documentation.pdf" your documentation, the group that is working on the library documentation will find it difficult to understand which function you are referring to. Change the name.

In the "Usage examples" section you also need to show the provided output and the validation with XRotor.

CrescenzoVisone committed:
Changed name of documentation .pdf file and added suggested corrections in the "usage examples" section

CrescenzoVisone commented:
> @CrescenzoVisone
> If you name "documentation.pdf" your documentation, the group that is working on the library documentation will find it difficult to understand which function you are referring to. Change the name.
> 
> In the "Usage examples" section you also need to show the provided output and the validation with XRotor.

Thank you for your suggestions @EttoreSaetta 
I have changed the name and added the output from XRotor in order to demonstrate the validity of the script. Is it ok to show the drag polar or did you want me to only give comparison values?

EttoreSaetta commented:
> I have changed the name and added the output from XRotor in order to demonstrate the validity of the script. Is it ok to show the drag polar or did you want me to only give comparison values?

@CrescenzoVisone It is better to show the comparison between the two curves in the same plot.

CrescenzoVisone committed:
Added corrections to doc file

CrescenzoVisone commented:
> > I have changed the name and added the output from XRotor in order to demonstrate the validity of the script. Is it ok to show the drag polar or did you want me to only give comparison values?
> 
> @CrescenzoVisone It is better to show the comparison between the two curves in the same plot.

@EttoreSaetta done!
